### PR TITLE
AO3-6393 AO3-6130 Prevent same language translations and ensure translations are deleted with original post

### PR DESCRIPTION
**Pull request by Cesium-Ice**
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6393 
https://otwarchive.atlassian.net/browse/AO3-6130

## Purpose

What does this PR do?
Prevent creation of a "translation" post in the same language as the original post, and ensure translations are deleted with the original post is deleted.

## Testing Instructions

How can the Archive's QA team verify that this is working as you intended?

If you have a Jira account with access, please update or comment on the issue
with any new or missing testing instructions instead.

## References

Are there other relevant issues/pull requests/mailing list discussions?

## Credit

What name and pronouns should we use to credit you in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?
Cesium-Ice, she/her
If you have a Jira account, please include the same name in the "Full name"
field on your Jira profile, so we can assign you the issues you're working on.

*Please note that if you do not fill in this section, we will use your GitHub account name and
they/them pronouns.*

### CODE REVIEW

**Review by sarken — CHANGES_REQUESTED** [review]
I think it would be better to handle this in the AdminPost model, like [the validation for making sure the translated post exists](https://github.com/otwcode/otwarchive/blob/0584883f846478b0a6f2eac9d6687d018c32fe02/app/models/admin_post.rb#L69-L73).

File: app/models/admin_post.rb [hunk 1]
```diff
@@ -11,8 +11,8 @@ class AdminPost < ApplicationRecord
   }, _suffix: :comments
 
   belongs_to :language
-  belongs_to :translated_post, class_name: 'AdminPost'
-  has_many :translations, class_name: 'AdminPost', foreign_key: 'translated_post_id'
+  belongs_to :translated_post, class_name: "AdminPost"
+  has_many :translations, class_name: "AdminPost", foreign_key: "translated_post_id"
```
[the remark this answers is not in the document]
Comment (Cesium-Ice) [reply app/models/admin_post.rb:15]: Relevant issue in Jira: https://otwarchive.atlassian.net/browse/AO3-6130

File: app/models/admin_post.rb [hunk 2]
```diff
@@ -68,10 +70,16 @@ def tag_list=(list)
 
   def translated_post_must_exist
     if translated_post_id.present? && AdminPost.find_by(id: translated_post_id).nil?
-      errors.add(:translated_post_id, 'does not exist')
+      errors.add(:translated_post_id, "does not exist")
     end
   end
 
+  def translated_post_language_must_differ
+    return unless translated_post_id.present? && !AdminPost.find_by(id: translated_post_id).nil? && self.language_id == AdminPost.find_by(id: translated_post_id).language_id
```
Comment (sarken) [inline app/models/admin_post.rb:78]: I think we can make this a little easier to read by breaking it up across two lines and simplifying it:
```
return unless translated_post.present?
return unless translated_post.language == language
```

File: app/models/admin_post.rb [hunk 3]
```diff
@@ -11,8 +11,8 @@ class AdminPost < ApplicationRecord
   }, _suffix: :comments
 
   belongs_to :language
-  belongs_to :translated_post, class_name: 'AdminPost'
-  has_many :translations, class_name: 'AdminPost', foreign_key: 'translated_post_id'
+  belongs_to :translated_post, class_name: "AdminPost"
+  has_many :translations, class_name: "AdminPost", foreign_key: "translated_post_id", dependent: :destroy
```
Comment (sarken) [inline app/models/admin_post.rb:15]: Could you add a test for this? I'm sure it works fine, but a test will let get our attention if we ever mistakenly remove `dependent: :destroy`. 😆 

The best thing to do is probably copy the existing `"deletes post and redirects without notice"` example and modify it to include a translation, e.g., `let!(:translation) { create(:admin_post, language: create(:language), translated_post: post) }`. That way, we'll have tests for deleting both posts with translations and without them.

**Review by sarken — CHANGES_REQUESTED** [review]

File: features/step_definitions/admin_steps.rb [hunk 4]
```diff
@@ -279,6 +279,19 @@
   click_button("Post")
 end
 
+When /^I make a same language translation of an admin post( with tags "(.*?)")?$/ do |tags|
```
[the remark this answers is not in the document]
Comment (Cesium-Ice) [reply features/step_definitions/admin_steps.rb:282]: removed features and added specs to test same language translation validation

File: app/models/admin_post.rb (hunk 3 above)
Comment (Cesium-Ice) [reply app/models/admin_post.rb:15]: done

File: spec/controllers/admin_posts_controller_spec.rb [hunk 5]
```diff
@@ -183,6 +204,18 @@
             expect { post.reload }.to raise_exception(ActiveRecord::RecordNotFound)
             it_redirects_to(admin_posts_path)
           end
+          context "with translated post" do
```
Comment (sarken) [inline spec/controllers/admin_posts_controller_spec.rb:207]: Can you add a blank line above this?

**Review by sarken — APPROVED** [review]

**Pull request merged**